### ISSUE TITLE
profiler trace_steps is not match global step in distribution training

### ISSUE THREAD

**Issue by PaulChongPeng** [tensorflow/tensorflow#21782]
### System information

== cat /etc/issue ===============================================
Linux gpu0198 3.10.0-229.el7.x86_64 #1 SMP Fri Mar 6 11:36:42 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
VERSION="7 (Core)"
VERSION_ID="7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

== are we in docker =============================================
Yes

== compiler =====================================================
c++ (GCC) 4.8.5 20150623 (Red Hat 4.8.5-28)
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

== uname -a =====================================================
Linux gpu0198 3.10.0-229.el7.x86_64 #1 SMP Fri Mar 6 11:36:42 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

== check pips ===================================================
numpy                              1.14.5
protobuf                           3.6.1
tensorflow-gpu                     1.10.0

== check for virtualenv =========================================
True

== tensorflow import ============================================
tf.VERSION = 1.10.0
tf.GIT_VERSION = v1.10.0-0-g656e7a2b34
tf.COMPILER_VERSION = v1.10.0-0-g656e7a2b34
Sanity check: array([1], dtype=int32)

== env ==========================================================
LD_LIBRARY_PATH :/usr/local/nvidia/cpu_lib:/nodemanager/lib/native:/usr/local/cuda/lib64:/usr/local/cuda/extras/CUPTI/lib64:/usr/local/java//jre/lib/amd64/server:/nodemanager/lib/native
DYLD_LIBRARY_PATH is unset

== nvidia-smi ===================================================
Wed Aug 22 03:12:21 2018
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 390.46                 Driver Version: 390.46                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-PCIE...  On   | 00000000:86:00.0 Off |                    0 |
| N/A   28C    P0    38W / 250W |      0MiB / 16160MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+

== cuda libs  ===================================================
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudart.so.9.0.176
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudart_static.a 

### Describe the problem

When I use tensorflow profiler to get the timeline for distribution training, I found the trace_steps didn't match the global step and it result in useless timeline.
I check the profile_context.py and find that the profile step is not related to the global step.

distribution config:
ps 2
worker 8
1 gpu per worker

### Source code / logs
sync code:
with tf.contrib.tfprof.ProfileContext(profile_dir=FLAGS.train_dir,
                                                trace_steps=range(100, 200,3),
                                                dump_steps=[200],
                                                enabled=FLAGS.enable_profile,
                                                debug=FLAGS.enable_profile_debug) as pctx:
...
...
                optimizer = tf.train.SyncReplicasOptimizer(
                    opt=optimizer,
                    replicas_to_aggregate=FLAGS.replicas_to_aggregate,
                    total_num_replicas=worker_replicas,
                    variable_averages=variable_averages,
                    variables_to_average=moving_average_variables)

sync train log :
worker 0:
debug: tracing step: 148
debug: tracing step: 151
INFO:tensorflow:global step 19: loss = 2.1377 (8.300 sec/step)
debug: tracing step: 154
debug: tracing step: 157

async code:
with tf.contrib.tfprof.ProfileContext(profile_dir=FLAGS.train_dir,
                                                trace_steps=range(800, 900),
                                                dump_steps=[200],
                                                enabled=FLAGS.enable_profile,
                                                debug=FLAGS.enable_profile_debug) as pctx:

async train log:
worker 0:
debug: tracing step: 838
INFO:tensorflow:global step 2198: loss = 2.0687 (6.361 sec/step)
debug: tracing step: 839

worker 7:
debug: tracing step: 825
INFO:tensorflow:global step 3137: loss = 2.5605 (7.343 sec/step)
debug: tracing step: 826

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by PaulChongPeng** [comment]
### System information
- **Have I written custom code ：Yes
- **OS Platform and Distribution: docker Centos7  with v100
- **Mobile device if the issue happens on mobile device: No
- **TensorFlow installed from: binary
- **TensorFlow version :1.10.0
- **Python version: 2.7
- **CUDA/cuDNN version: cuda-9.0 and cudnn 7

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by poxvoculi** [comment]
@asimshankar: Who maintains this tool?

**Comment by asimshankar** [comment]
@zhangyaobit : Mind taking a look?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zhangyaobit: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by wt-huang** [comment]
@PaulChongPeng The trace_steps used in timeline is intended to capture just a single step when program makes a single call to `session.run()`.  Each `session.run()` generates `run_metadata` protobuf which is overwritten every time. `global_step` doesn't get overwritten hence the difference as seen above.

**Comment by PaulChongPeng** [comment]
So tensorflow profiler doesn't support to get the timeline for a global step when training distribution? 
How can I get the the timeline for a global step to check my distribution training performance bottleneck？

**Comment by wt-huang** [comment]
global_step should be fine and it is not affected. You can access and get the timeline for global_step using tensorflow profiler. Make sure to include tf.contrib.framework.create_or_get_global_step() or tf.train.create_or_get_global_step() in your code.

**Comment by PaulChongPeng** [comment]
I checked my code ,and confirmed  that I had used tf.train.create_global_step() in my training code.
I checked tensorflow profiler code again:
 _profiled_run function in tensorflow/tensorflow/python/profiler/profile_context.py use profile_context._new_step() and profile_context._should_trace(self, step, graph, fetches) to check whether should trace or not.

  def _new_step(self):
    acquired = self._lock.acquire(False)
    yield (self._step, acquired)
    self._step += 1
    self._trace_next_step = False
    self._dump_next_step = False
    if acquired:
      self._lock.release()

  def _should_trace(self, step, graph, fetches):
    """Whether should do tracing at current step."""
    if self._traced_steps > MAX_TRACED_STEPS:
      return False
    # Check user-set tracing steps.
    if step in self._trace_steps or self._trace_next_step:
      self._traced_steps += 1
      return True

The step  which is used to check whether should trace or not is not related to the global step.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @wt-huang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tilakrayal** [comment]
@PaulChongPeng 
Is this still an issue?
Could you update TensorFlow to the latest stable version v2.9 and let us know if you are facing the same error. Thanks!
